File: brace_check.js
const fs = require('fs');
const text = fs.readFileSync('lib/main.dart', 'utf8');
const stack = [];
let state = 'code';
let quote = null;
let i = 0;
while (i < text.length) {
  const ch = text[i];
  const nxt = i + 1 < text.length ? text[i + 1] : '';
  if (state === 'code') {
    if (ch === '/' && nxt === '/') {
      state = 'linecomment';
      i += 2;
      continue;
    }
    if (ch === '/' && nxt === '*') {
      state = 'blockcomment';
      i += 2;
      continue;
    }
    if (ch === '"' || ch === "'" || ch === '`') {
      state = 'string';
      quote = ch;
      i += 1;
      continue;
    }
    if (ch === '{') stack.push(i);
    else if (ch === '}') {
      if (!stack.length) {
        console.log('UNMATCHED CLOSE', i);
        break;
      }
      stack.pop();
    }
  } else if (state === 'linecomment') {
    if (ch === '\n') state = 'code';
  } else if (state === 'blockcomment') {
    if (ch === '*' && nxt === '/') {
      state = 'code';
      i += 2;
      continue;
    }
  } else if (state === 'string') {
    if (ch === '\\') {
      i += 2;
      continue;
    }
    if (ch === quote) state = 'code';
  }
  i += 1;
}
console.log('UNMATCHED OPEN', stack.length);
for (const pos of stack.slice(-10)) {
  const before = text.slice(0, pos).split('\n');
  console.log('line', before.length, 'col', before[before.length - 1].length + 1, JSON.stringify(before[before.length - 1].slice(-80)));
}

File: brace_check.py
import pathlib

text = pathlib.Path('lib/main.dart').read_text(encoding='utf-8')
stack = []
state = 'code'
quote = None
i = 0
while i < len(text):
    ch = text[i]
    nxt = text[i + 1] if i + 1 < len(text) else ''
    if state == 'code':
        if ch == '/' and nxt == '/':
            state = 'linecomment'
            i += 2
            continue
        if ch == '/' and nxt == '*':
            state = 'blockcomment'
            i += 2
            continue
        if ch in ('"', "'", '`'):
            state = 'string'
            quote = ch
            i += 1
            continue
        if ch == '{':
            stack.append(i)
        elif ch == '}':
            if not stack:
                print('UNMATCHED CLOSE', i)
                break
            stack.pop()
    elif state == 'linecomment':
        if ch == '\n':
            state = 'code'
    elif state == 'blockcomment':
        if ch == '*' and nxt == '/':
            state = 'code'
            i += 2
            continue
    elif state == 'string':
        if ch == '\\':
            i += 2
            continue
        if ch == quote:
            state = 'code'
    i += 1
print('UNMATCHED OPEN', len(stack))
for pos in stack[-10:]:
    prefix = text[:pos].split('\n')
    print('line', len(prefix), 'col', len(prefix[-1]) + 1, repr(prefix[-1][-80:]))
